### PR TITLE
Add torch.nn.GELU for GELU activation

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary: 

1. Add torch.nn.GELU for GELU activation
2. Fix fp16 failures for torch.nn.functional.gelu on CUDA
3. Add infinitely differentiable backward implementation to support higher order gradient.
4. Change gelu_cpu kernel to use TensorIterator.

Differential Revision: D18240946

### CODE REVIEW

**Comment by BIT-silence** [comment]
link to https://github.com/pytorch/pytorch/pull/28944

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit dc2112f:
* 0/1 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=dc2112ff93e4fbd1f518a7c2cc554fde60652990).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by SsnL** [comment]
This does much more than #28947 and the description of this PR. Maybe we can update the description and title of this?

**Review by soumith — APPROVED** [review]
lgtm!

**Comment by BramVanroy** [comment]
I'm happy to see these lines indicating that another kernel might get implemented, too, but I wonder how you are going to implement this as an interface.

https://github.com/pytorch/pytorch/blob/53c2d01ebc0e40c8982ccd2e9cb4113e1d0b5bc1/aten/src/ATen/native/cpu/Activation.cpp#L138-L140

Will it be a separate function/module, or rather a parameter setting?

**Comment by BIT-silence** [comment]
> I'm happy to see these lines indicating that another kernel might get implemented, too, but I wonder how you are going to implement this as an interface. I've had great results with that implementation as a custom module.
> 
> https://github.com/pytorch/pytorch/blob/53c2d01ebc0e40c8982ccd2e9cb4113e1d0b5bc1/aten/src/ATen/native/cpu/Activation.cpp#L138-L140
> 
> Will it be a separate function/module, or rather a parameter setting?

Maybe we will add it with a parameter in gelu later. Currently the reason we didn't do that is actually that approximation will not provide a better performance compare to current implementation with MKL functions. In my testing, the tanh approximation's performance rely on the performance of tanh. The eigen which is the backend of TensorFlow provides a fast approximation of tanh here. Without that tanh implementation, I think it is not very necessary to add the approximation for gelu here.
https://github.com/eigenteam/eigen-git-mirror/blob/8e409c71423fc070239cad72afdbd973727b5e0f/Eigen/src/Core/MathFunctionsImpl.h#L26

**Comment by calclavia** [comment]
Is this released on torch 1.3.1 because it wasn't fixed in that version.

**Comment by yf225** [comment]
@calclavia This PR is not in PyTorch 1.3.1. It will be in our upcoming release 1.4.0.
